modules/priceFloors.js

```diff
@@ -0,0 +1,642 @@
+ * a "*" catch-all match
+ * Returns array of Tuple [exact match, catch all] for each field in rules file
+ */
+function enumeratePossibleFieldValues(floorFields, bidObject, responseObject) {
+  // generate combination of all exact matches and catch all for each field type
+  return floorFields.reduce((accum, field) => {
+    let exactMatch = fieldMatchingFunctions[field](bidObject, responseObject);
+    // storing exact matches as lowerCase since we want to compare case insensitively
+    accum.push(exactMatch === '*' ? ['*'] : [exactMatch.toLowerCase(), '*']);
+    return accum;
+  }, []);
+};
+
+/**
+ * @summary get's the first matching floor based on context provided.
+ * Generates all possible rule matches and picks the first matching one.
+ */
+export function getFirstMatchingFloor(floorData, bidObject, responseObject = {}) {
+  let fieldValues = enumeratePossibleFieldValues(utils.deepAccess(floorData, 'schema.fields') || [], bidObject, responseObject);
+  if (!fieldValues.length) return { matchingFloor: floorData.default };
+
+  // look to see iof a request for this context was made already
+  let matchingInput = fieldValues.map(field => field[0]).join('-');
+  // if we already have gotten the matching rule from this matching input then use it! No need to look again
+  let previousMatch = utils.deepAccess(floorData, `matchingInputs.${matchingInput}`);
+  if (previousMatch) {
+    return previousMatch;
+  }
+  let allPossibleMatches = generatePossibleEnumerations(fieldValues, utils.deepAccess(floorData, 'schema.delimiter') || '|');
+  let matchingRule = find(allPossibleMatches, hashValue => floorData.values.hasOwnProperty(hashValue));
+
+  let matchingData = {
+    matchingFloor: floorData.values[matchingRule] || floorData.default,
+    matchingData: allPossibleMatches[0], // the first possible match is an "exact" so contains all data relevant for anlaytics adapters
+    matchingRule
+  };
+  // save for later lookup if needed
+  utils.deepSetValue(floorData, `matchingInputs.${matchingInput}`, {...matchingData});
+  return matchingData
+};
+
+/**
+ * @summary Generates all possible rule hash's based on input array of array's
+ * The generated list is of all possible key matches based on fields input
+ * The list is sorted by least amount of * in rule to most with left most fields taking precedence
+ */
+function generatePossibleEnumerations(arrayOfFields, delimiter) {
+  return arrayOfFields.reduce((accum, currentVal) => {
+    let ret = [];
+    accum.map(obj => {
+      currentVal.map(obj1 => {
+        ret.push(obj + delimiter + obj1)
+      });
+    });
+    return ret;
+  }).sort((left, right) => (left.split('*').length - 1) - (right.split('*').length - 1));
+};
+
+/**
+ * @summary If a the input bidder has a registered cpmadjustment it returns the input CPM after being adjusted
+ */
+export function getBiddersCpmAdjustment(bidderName, inputCpm) {
```

Comment:
Will the value of `inputCpm` or the returned value from the adjustmentFunction ever not be a number? Not sure if `parseFloat` is needed

Reply:
`adjustmentFunction` is something publisher's set on their page.

Most often it is simply something like:

`return cpm * 0.8`

So in that case it is _probably_ always a number.

But who knows, they could possibly be doing some of their own rounding and thus returning a string maybe?